lib/screens/v2/candidate_generator/widgets/reconstructed_alert.ex

```diff
@@ -26,26 +38,142 @@ defmodule Screens.V2.CandidateGenerator.Widgets.ReconstructedAlert do
+            common_parameters
+          ) ++
+            create_alert_instances(downstream_disruptions, false, common_parameters) ++
+            create_alert_instances(moderate_delays, false, common_parameters)
+
+        Enum.any?(downstream_disruptions) ->
+          fullscreen_alerts =
+            find_closest_downstream_alerts(
+              downstream_disruptions,
+              stop_id,
+              location_context.stop_sequences
+            )
+
+          flex_zone_alerts = downstream_disruptions -- fullscreen_alerts
+
+          create_alert_instances(fullscreen_alerts, true, common_parameters) ++
+            create_alert_instances(flex_zone_alerts, false, common_parameters) ++
+            create_alert_instances(moderate_delays, false, common_parameters)
+
+        true ->
+          create_alert_instances(moderate_delays, true, common_parameters)
+      end
     else
       :error -> []
     end
   end
 
+  defp create_alert_instances(
+         alerts,
+         is_full_screen,
+         config: config,
+         location_context: location_context,
+         fetch_stop_name_fn: fetch_stop_name_fn,
+         is_terminal_station: is_terminal_station,
+         now: now
+       ) do
+    Enum.map(alerts, fn alert ->
+      %ReconstructedAlert{
+        screen: config,
+        alert: alert,
+        now: now,
+        location_context: location_context,
+        informed_stations_string: get_stations(alert, fetch_stop_name_fn),
+        is_terminal_station: is_terminal_station,
+        is_full_screen: is_full_screen
+      }
+    end)
+  end
+
+  defp find_closest_downstream_alerts(alerts, stop_id, stop_sequences) do
+    home_stop_distance_map = build_distance_map(stop_id, stop_sequences)
+    # Map each alert with its distance from home.
+    alerts
+    |> Enum.map(fn %{informed_entities: ies} = alert ->
+      distance =
+        ies
+        |> Enum.filter(&String.starts_with?(&1.stop, "place-"))
+        |> Enum.map(&get_distance(home_stop_distance_map, &1))
+        |> Enum.min()
+
+      {alert, distance}
+    end)
+    |> Enum.group_by(&elem(&1, 1), &elem(&1, 0))
+    |> Enum.sort_by(&elem(&1, 0))
+    # The first item will be all alerts with the shortest distance.
+    |> List.first()
+    |> elem(1)
+  end
+
+  defp build_distance_map(home_stop_id, stop_sequences) do
+    Enum.reduce(stop_sequences, %{}, fn stop_sequence, distances_by_stop ->
+      stop_sequence
+      # Index each element by its distance from home_stop_id. For example if home_stop_id is at position 2, then indices would start at -2.
+      |> Enum.with_index(-Enum.find_index(stop_sequence, &(&1 == home_stop_id)))
+      # Convert negative distances to positive, and put into a map.
+      |> Map.new(fn {stop, d} -> {stop, abs(d)} end)
+      # Merge with the distances recorded from previous stop sequences.
+      # If a stop already has a distance recorded, we use the smaller of the two. <-- *** Unsure if this is always what we'd want to do! ***
```

Comment:
I can't think of a situation where this would happen. Two stations should be the same number of stops apart in every stop sequence. Do you have an example? If not, I guess I'd just say go with d1?